extensions/amp-form/0.1/amp-form.js

```diff
@@ -721,42 +724,78 @@ function isDisabled_(element) {
 
 
 /**
- * Installs submission handler on all forms in the document.
- * @param {!Window} win
+ * Bootstraps the amp-form elements
  */
-function installSubmissionHandlers(win) {
-  onDocumentReady(win.document, doc => {
-    toArray(doc.forms).forEach((form, index) => {
+export class AmpFormService {
+  /**
+   * @param  {!../../../src/service/ampdoc-impl.AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+
+    /** @const {!Document|!ShadowRoot} */
+    this.rootNode_ = ampdoc.getRootNode();
+
+    /** @const {boolean} */
+    this.isShadow_ = isShadowRoot(this.rootNode_);
+
+    this.initialize_();
+  }
+
+  /**
+   * Install the amp-form CSS and handlers on the ampdoc.
+   */
+  initialize_() {
+    if (this.isShadow_) {
+      const root = /** @type {!ShadowRoot} */ (this.rootNode_);
+      installStylesForShadowRoot(root, CSS);
+      this.install_();
+    } else {
+      const root = /** @type {!Document} */ (this.rootNode_);
+      installStyles(root, CSS, () => this.install_());
+    }
+  }
+
+  /**
+   * Install the appropriate handlers depending on the type of ampdoc.
+   */
+  install_() {
+    if (this.isShadow_) {
+      this.installSubmissionHandlers_(
+          this.rootNode_.getElementsByTagName('form'));
+    } else {
+      const doc = /** @type {!Document} */ (this.rootNode_);
+      onDocumentReady(doc, () => {
+        this.installSubmissionHandlers_(doc.forms);
+      });
+      this.installGlobalEventListener_(doc);
+    }
+  }
+
+  /**
+   * Installs submission handler on all forms in the document.
+   * @param {?IArrayLike} forms
+   * @previousValidityState
+   * @private
+   */
+  installSubmissionHandlers_(forms) {
+    toArray(forms).forEach((form, index) => {
```

Comment:
When you say "next" iteration, do you mean the ES6 `Symbol.iterator` or just use a for-loop?